### PR TITLE
add sqs provider

### PR DESCRIPTION
**Pull request by nicolasThal**
I will share the provider I have written to use swarrot with amazon sqs.

This code is already in production on our platform.

As we are currently changing our queueing system, we didn't use a message publisher yet

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -19,12 +19,15 @@
         "phpunit/phpunit": "~4.7|~5.0",
         "php-amqplib/php-amqplib": "~2.1",
         "doctrine/common": "~2.3",
-        "doctrine/dbal": "~2.2"
+        "doctrine/dbal": "~2.2",
+        "aws/aws-sdk-php": "2.*",
+        "phake/phake": "^2.3"
```
Comment (odolbeau) [inline composer.json]: Is it possible to rewrite tests to use prophecy?

File: src/Swarrot/Broker/MessageProvider/SqsMessageProvider.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+     * @param SqsClient                  $channel
+     * @param string                     $queueName
+     * @param MessageCacheInterface|null $cache
+     */
+    public function __construct(SqsClient $channel, $queueName, MessageCacheInterface $cache = null)
+    {
+        $this->channel = $channel;
+        $this->queueName = $queueName;
+        $this->cache = $cache?:new PrefetchMessageCache();
+        $this->prefetch = 9;
```
Comment (odolbeau) [inline src/Swarrot/Broker/MessageProvider/SqsMessageProvider.php]: Could you add this as a constructor argument?

File: src/Swarrot/Broker/MessageProvider/SqsMessageProvider.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    public function get()
+    {
+        if ($message = $this->cache->pop($this->getQueueName())) {
+            return $message;
+        }
+
+        $result = $this->channel->receiveMessage([
+            'QueueUrl' => $this->getQueueName(),
+            'MaxNumberOfMessages' => $this->prefetch,
+            'WaitTimeSeconds' => 5,
```
Comment (odolbeau) [inline src/Swarrot/Broker/MessageProvider/SqsMessageProvider.php]: Could you add this as a constructor argument?

File: src/Swarrot/Broker/MessageProvider/SqsMessageProvider.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    public function __construct(SqsClient $channel, $queueName, MessageCacheInterface $cache = null)
+    {
+        $this->channel = $channel;
+        $this->queueName = $queueName;
+        $this->cache = $cache?:new PrefetchMessageCache();
+        $this->prefetch = 9;
+    }
+
+    /**
+     * get.
```
Comment (odolbeau) [inline src/Swarrot/Broker/MessageProvider/SqsMessageProvider.php]: inheritdoc (same below)

**Comment by odolbeau** [comment]
Thanks! <3

**Comment by Taluu** [comment]
The MessageCache classes are not within the scope of this PR IMO.

**Comment by odolbeau** [comment]
@Taluu I think it is. RMQ also supports prefetch. The `MessageCacheInterface` could be used in several providers including the RMQ one.

**Comment by Taluu** [comment]
Yes, but IMO, it should be in a different PR. I'm quite the partisan of "one pr does one thing" : here, it should just add the providers (and producers) for sqs. Another PR could add the prefetch. :}

**Comment by odolbeau** [comment]
Not really a problem in this case. A dedicated PR for the `MessageCacheInterface` doesn't make sense as it's not usable without this provider.
IMO that's a prerequisite which is shipped in the same PR and it's fine like this. Furthermore, it's only 2 small classes. ;)

File: composer.json [hunk 5]
```diff
@@ -16,15 +16,17 @@
         "psr/log": "~1.0"
     },
     "require-dev": {
-        "phpunit/phpunit": "~4.7|~5.0",
+        "phpunit/phpunit": "~4.8.27|~5.0",
```
Comment (Taluu) [inline composer.json]: why ? :o

Comment (stof) [inline composer.json]: probbaly to make tests running with ``--prefer-lowest`` working when using the PHPUnit version installed through composer

File: composer.json [hunk 6]
```diff
@@ -16,15 +16,17 @@
...
         "php-amqplib/php-amqplib": "~2.1",
         "doctrine/common": "~2.3",
-        "doctrine/dbal": "~2.2"
+        "doctrine/dbal": "~2.2",
+        "aws/aws-sdk-php": "~2.8"
```
Comment (stof) [inline composer.json]: wouldn't it be better to build it against version 3 of the SDK ?

File: src/Swarrot/Broker/MessageProvider/SqsMessageProvider.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * Class SqsMessageProvider.
+ */
+class SqsMessageProvider implements MessageProviderInterface
+{
+    protected $cache;
+    protected $channel;
+    protected $prefetch;
+    protected $queueName;
```
Comment (stof) [inline src/Swarrot/Broker/MessageProvider/SqsMessageProvider.php]: @odolbeau I suggest using private properties. It makes it much easier to manage backward compatibility

File: src/Swarrot/Driver/MessageCacheInterface.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+use Swarrot\Broker\Message;
+
+/**
+ * Interface MessageCacheInterface.
+ */
+interface MessageCacheInterface
+{
+    /**
+     * Pushes a $message to the end of the cache.
```
Comment (stof) [inline src/Swarrot/Driver/MessageCacheInterface.php]: I would say ``Pushes a message``. No need to use a variable name in the phpdoc

File: composer.json (hunk 6 above)
Comment (nicolasThal) [inline composer.json]: The sdk requires a version of php > 5.5 on version 3. 
https://github.com/aws/aws-sdk-php/blob/3.0.0/composer.json#L19

File: composer.json (hunk 5 above)
Comment (nicolasThal) [inline composer.json]: Yes, I had some issue with phpdocumentor.

https://github.com/phpDocumentor/ReflectionDocBlock/issues/80

**Comment by nicolasThal** [comment]
@odolbeau I have updated the pr

File: composer.json [hunk 9]
```diff
@@ -16,15 +16,17 @@
...
     },
     "suggest": {
         "pecl-amqp": "*",
         "php-amqplib/php-amqplib": "~2.1",
-        "symfony/console": "~2.4|~3.0"
+        "symfony/console": "~2.4|~3.0",
+        "aws/aws-sdk-php": ">=2.8"
```
Comment (odolbeau) [inline composer.json]: I don't know this SDK at all.
Is it possible to have `^2.8 || ^3.0` maybe?

**Comment by odolbeau** [comment]
Looks good right now. :)
Can you rebase your branch and change the aws-sdk dependency please? :)

**Comment by nicolasThal** [comment]
@odolbeau i have rebase the branch.

For the sdk version, I perfer to use the version 2, the 3 has a lot of bc breaks.

**Comment by Taluu** [comment]
with your `">= 2.8"` constraint, it * will * install the 3.0 if it is available. So you should at least change this to `^2.8`, (which means `>=2.8, < 3.0`). But you should also consider of trying to make something compatible with both `2.8` and `3.0` imo

**Comment by nicolasThal** [comment]
@Taluu I have already put `^2.8` for the version.

As the method to get the element from the sqs queue are not the same in the sdk, I will be a lot of work to make the code compatible with both sdk version

**Comment by odolbeau** [comment]
Looks good for me!
Is it ok for you @Taluu & @stof?

**Comment by odolbeau** [comment]
Looks like it's OK for everyone.

Again, thanks for this awesome contribution @nicolasThal! :)

**Pull request merged**